keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,151 @@
+    y_true = tf.random.uniform((5, 10, 5), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    y_pred = tf.random.uniform((5, 10, 4), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    loss = GIoULoss(bounding_box_format = "xyWH")
+    loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.GIoULoss())
+    ```
+    """
+
+    def __init__(self, bounding_box_format, axis=None, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.axis = axis
+
+    def _compute_giou(self, boxes1, boxes2):
+        boxes1_rank = len(boxes1.shape)
+        boxes2_rank = len(boxes2.shape)
+
+        if (
+            boxes1_rank != boxes2_rank
+            or boxes1_rank not in [2, 3]
+            or boxes2_rank not in [2, 3]
+        ):
+            raise ValueError(
+                "`GIoULoss` expects both boxes to be batched, or both "
+                f"boxes to be unbatched.  Received `len(boxes1.shape)`={boxes1_rank}, "
+                f"`len(boxes2.shape)`={boxes2_rank}.  Expected either `len(boxes1.shape)`=2 AND "
+                "`len(boxes2.shape)`=2, or `len(boxes1.shape)`=3 AND `len(boxes2.shape)`=3."
+            )
+
+        if self.bounding_box_format.startswith("rel"):
+            target = "rel_yxyx"
+        else:
+            target = "yxyx"
+
+        boxes1 = bounding_box.convert_format(
+            boxes1, source=self.bounding_box_format, target=target
+        )
+
+        boxes2 = bounding_box.convert_format(
+            boxes2, source=self.bounding_box_format, target=target
+        )
+
+        def compute_giou_for_batch(boxes):
+            boxes1, boxes2 = boxes
+            zero = tf.convert_to_tensor(0.0, boxes1.dtype)
+            boxes1_ymin, boxes1_xmin, boxes1_ymax, boxes1_xmax = tf.unstack(
+                boxes1[..., :4], 4, axis=-1
+            )
+            boxes2_ymin, boxes2_xmin, boxes2_ymax, boxes2_xmax = tf.unstack(
+                boxes2[..., :4], 4, axis=-1
+            )
+            boxes1_width = tf.maximum(zero, boxes1_xmax - boxes1_xmin)
+            boxes1_height = tf.maximum(zero, boxes1_ymax - boxes1_ymin)
+            boxes2_width = tf.maximum(zero, boxes2_xmax - boxes2_xmin)
+            boxes2_height = tf.maximum(zero, boxes2_ymax - boxes2_ymin)
+            boxes1_area = boxes1_width * boxes1_height
+            boxes2_area = boxes2_width * boxes2_height
+            intersect_ymin = tf.maximum(boxes1_ymin, boxes2_ymin)
+            intersect_xmin = tf.maximum(boxes1_xmin, boxes2_xmin)
+            intersect_ymax = tf.minimum(boxes1_ymax, boxes2_ymax)
+            intersect_xmax = tf.minimum(boxes1_xmax, boxes2_xmax)
+            intersect_width = tf.maximum(zero, intersect_xmax - intersect_xmin)
+            intersect_height = tf.maximum(zero, intersect_ymax - intersect_ymin)
+            intersect_area = intersect_width * intersect_height
```

Comment:
please re-use bbox.compute_iou

Reply:
This will be impossible without making the API noisy and changing `compute_iou`. In my opinion, this is not ideal for a user-facing API.

Alternatively, using `compute_iou` for what it already has will reduce the lines of code by only ~8 lines and make the code considerably more confusing.

Reply:
besides the benefit of reducing code complexity, if we find bugs for computing iou, it'd be easier to track where this needs to be fixed as well

Reply:
I'm not sure I understand why we need to change `compute_iou` to make it work here, can you elaborate?

Reply:
> I'm not sure I understand why we need to change compute_iou to make it work here, can you elaborate?

The area of the enclosing box is also required for GIoU. So the code starting line 116 is added to `compute_iou` to form `compute_giou`.

An alternative if we had to reuse `compute_iou` was to update the API to take in a string indicating whether IoU or GIoU is being computed (as in the addons implementation). But since this is a user-facing API, I decided to let it be after consulting Luke.

Reply:
I mean from line 101 to line 112 can be replaced by `compute_iou`, while the giou part can still be manually computed here, correct?

Reply:
This might be somewhat fixed now that I am using utils from `compute_iou`